techlibs/nexus/synth_nexus.cc

```diff
@@ -0,0 +1,376 @@
+				continue;
+			}
+			break;
+		}
+		extra_args(args, argidx, design);
+
+		if (!design->full_selection())
+			log_cmd_error("This command only operates on fully selected designs!\n");
+
+		if (abc9 && retime)
+				log_cmd_error("-retime option not currently compatible with -abc9!\n");
+
+		log_header(design, "Executing SYNTH_NEXUS pass.\n");
+		log_push();
+
+		run_script(design, run_from, run_to);
+
+		log_pop();
+	}
+
+	void script() override
+	{
+
+		if (family != "lifcl" && family != "lfd2nx")
+			log_cmd_error("Invalid Nexus -family setting: '%s'.\n", family.c_str());
+
+		if (check_label("begin"))
+		{
+			run("read_verilog -lib -specify +/nexus/cells_sim.v +/nexus/cells_xtra.v");
+			run(stringf("hierarchy -check %s", help_mode ? "-top <top>" : top_opt.c_str()));
+		}
+
+		if (check_label("coarse"))
+		{
+			run("proc");
+			if (flatten || help_mode)
+				run("flatten");
+			run("tribuf -logic");
+			run("deminout");
+			run("opt_expr");
+			run("opt_clean");
+			run("check");
+			run("opt -nodffe -nosdff");
+			run("fsm");
+			run("opt");
+			run("wreduce");
+			run("peepopt");
+			run("opt_clean");
+			run("share");
+			run("techmap -map +/cmp2lut.v -D LUT_WIDTH=4");
+			run("opt_expr");
+			run("opt_clean");
+
+			run("alumacc");
+			run("opt");
+			run("memory -nomap");
+			run("opt_clean");
+		}
+
+		if (!nobram && check_label("map_bram", "(skip if -nobram)"))
+		{
+			run("memory_bram -rules +/nexus/brams.txt");
+			run("setundef -zero -params t:$__NX_PDP16K");
+			run("techmap -map +/nexus/brams_map.v");
+		}
+
+		if (!nolutram && check_label("map_lutram", "(skip if -nolutram)"))
+		{
+			run("memory_bram -rules +/nexus/lutrams.txt");
+			run("setundef -zero -params t:$__NEXUS_DPR16X4");
+			run("techmap -map +/nexus/lutrams_map.v");
+		}
+
+		if (check_label("map_ffram"))
+		{
+			run("opt -fast -mux_undef -undriven -fine");
+			run("memory_map -iattr -attr !ram_block -attr !rom_block -attr logic_block "
+			    "-attr syn_ramstyle=auto -attr syn_ramstyle=registers "
+			    "-attr syn_romstyle=auto -attr syn_romstyle=logic");
+			run("opt -undriven -fine");
+		}
+
+		if (check_label("map_gates"))
+		{
+			if (help_mode || !noiopad)
+				run("iopadmap -bits -outpad OB I:O -inpad IB O:I -toutpad $__NX_TOUTPAD OE:I:O -tinoutpad $__NX_TINOUTPAD OE:O:I:B A:top", "(skip if '-noiopad')");
+			if (noccu2)
+				run("techmap");
+			else
+				run("techmap -map +/techmap.v -map +/nexus/arith_map.v");
+			run("opt -fast");
+			if (retime || help_mode)
+				run("abc -dff -D 1", "(only if -retime)");
+		}
+
+		if (check_label("map_ffs"))
+		{
+			run("opt_clean");
+			std::string dfflegalize_args = " -cell $_DFF_?_ 01 -cell $_DFF_?P?_ r -cell $_SDFF_?P?_ r";
+			if (help_mode) {
+				dfflegalize_args += " [-cell $_DFFE_??_ 01 -cell $_DFFE_?P??_ r -cell $_SDFFE_?P??_ r]";
+			} else if (!nodffe) {
+				dfflegalize_args += " -cell $_DFFE_??_ 01 -cell $_DFFE_?P??_ r -cell $_SDFFE_?P??_ r";
+			}
+			dfflegalize_args += " -cell $_DLATCH_?_ x";
```

Comment:
My understanding from @mwkmwkmwk is that wire loop latches are not useful from `dfflegalize`'s point of view.

Reply:
Ack, I was copying this from synth_ecp5 but it seems like that could be tidied up, too

Reply:
Just tried removing it for synth_ecp5 but that doesn't seem to work:

```
ERROR: FF latchp.$auto$simplemap.cc:581:simplemap_dlatch$354 (type $_DLATCH_P_) cannot be legalized: dlatch are not supported
```

Maybe I am missing something though?

Reply:
there are some plans to improve it, but until we have a better solution, `$_DLATCH_*` needs to be present here